SwiftCompilerSources/Sources/Optimizer/Utilities/LifetimeDependenceUtils.swift

```diff
@@ -0,0 +1,505 @@
+///
+///   %dependent = mark_dependence [nonescaping] %value on %value
+///
+/// Self-dependence is useful to ensure that derived values, including copies, do not escape the lifetime of the original value. Non-escapable function arguments are implicitly self-dependent, meaning that the argument's value does not escape the function body. Note that we do not insert a 'mark_dependence [nonescaping]' for function arguments because the caller must already represent the argument's dependence on some parent value. That parent value may not be the value directly passed to the argument. After inlining, an additional self-dependence on argument value would be overly strict.
+struct LifetimeDependence {
+  enum Scope {
+    /// A non-escapable or guaranteed argument whose scope is provided
+    /// by the caller and covers the entire function.
+    case caller(Argument)
+    /// An access scope.
+    case access(BeginAccessInst)
+    /// An owned value whose OSSA lifetime encloses nonescapable values
+    case owned(Value)
+    /// An singly-initialized address-only value, in which the entire
+    /// initialized region is a lifetime (as opposed to an individual
+    /// access). e.g. A value produced by @out.
+    case initialized(Value)
+
+    var parentValue: Value {
+      switch self {
+      case let .caller(argument): return argument
+      case let .access(beginAccess): return beginAccess
+      case let .owned(value): return value
+      case let .initialized(value): return value
+      }
+    }
+
+    func checkPrecondition() {
+      switch self {
+      case let .caller(argument):
+        precondition(!argument.type.isEscapable
+          || argument.ownership == .guaranteed,
+          "only non-escapable or guaranteed arguments have a caller scope")
+      case .access:
+        break
+      case let .owned(value):
+        precondition(value.ownership == .owned)
+      case let .initialized(value):
+        precondition(value.type.isAddress, "expected an address")
+      }
+    }
+    var description: String {
+      {
+        switch self {
+        case .caller: "Caller: "
+        case .access: "Access: "
+        case .owned: "Owned: "
+        case .initialized: "Initialized: "
+        }
+      }() + "\(parentValue)"
+    }
+  }
+  let scope: Scope
+  let dependentValue: Value
+  
+  var parentValue: Value { scope.parentValue }
+
+  var function: Function {
+    dependentValue.parentFunction // dependentValue can't be undef
+  }
+
+  var description: String {
+    return scope.description +
+      """
+      dependent: \(dependentValue)
+      """
+  }
+}
+
+extension LifetimeDependence {
+  /// Construct LifetimeDependence from mark_dependence [nonescaping]
+  ///
+  /// TODO: Add SIL verification that all mark_depedence [nonescaping]
+  /// have a valid LifetimeDependence.
+  init?(markDependenceInst: MarkDependenceInst, _ context: some Context) {
+    if markDependenceInst.escaping == .nonEscaping { return nil }
```

Comment:
`markDependenceInst.escaping != .nonEscaping`/`guard ... else`?

Reply:
I was following the no-double-negative rule. We can both be happy though
`    guard markDependenceInst.escaping == .escaping else { return nil }`

Reply:
Sorry, didn't mean to nitpick the style, I was just confused about the intent: From the comment (`Construct LifetimeDependence from mark_dependence [nonescaping]`), I was thinking that
```
mark_dependence [nonescaping]
```
would be required to init here, and that such a `markDependenceInst` would have `.escaping == .nonEscaping`, so bailing when `markDependenceInst.escaping == .nonEscaping` would be backwards.